probe-rs/src/vendor/st/sequences/stm32_armv7.rs

```diff
@@ -74,14 +82,21 @@ impl ArmDebugSequence for Stm32Armv7 {
         default_ap: &FullyQualifiedApAddress,
         _permissions: &crate::Permissions,
     ) -> Result<(), ArmError> {
-        let mut memory = interface.memory_interface(default_ap)?;
+        let mut saved_cr_value = self.saved_cr_value.lock().unwrap();
+        if saved_cr_value.is_some() {
+            return Ok(());
```

Comment:
Why dont you check the value and set it if its not whats expected?

Reply:
Not sure what you mean. The idea is to save the *original* value of the register, so I can only do that on the first invocation. I guess I could still run the rest of the function that flips the debug bits in the register and only skip saving the CR value, but those bits should all already be set from the first invocation, unless the program that's being debugged itself clears those bits while being traced. 

Does anyone actually ever call this function multiple times on the same `Stm32Armv7` instance?

Reply:
I'm actually not sure that it's a good idea to make the `Sequence` stateful. IIUC it's owned (through an `Arc`) by `Target`, not `Session`. It may be more logical to store the saved register value in the Session, although if so, I will need some guidance as to how best to do that.

Reply:
Ah NVM me, maybe I was tired.

Sequences are meant to be stateful and manufacturer specific. The ARM spec provided sequences are also stateful.
That's why the CR register should be stored in there :)

Reply:
What you potentially forget in this assumption here though is that the user could set those flags in the FW and then you do not read the register again with the current values, which might be necessary.

Reply:
I'm starting to think that the easier fix is simply to leave these flags enabled when the session is closed. Yes, that means that attaching the debugger has a persistent side effect, but it's at least a better side effect that we have right now.

Respective downsides:

1. Current code: if you attach a debugger, you cannot attach it again (or re-flush the device) once you detach.
2. Not clearing the flags: if you attach a debugger, your sleep modes will be less energy efficient once you detach.

The second downside seems less likely to be harmful, usually you would disconnect the probe and reset a device before leaving it running on battery. OTOH, reflashing after you debug the FW is very common.

If we continue with the approach of restoring the CR, there's still inevitably the possibility that firmware updates to those flags made while the debugger was attached may be lost. But if you think that's preferable, I can update the PR to refresh the stored CR value on every call to `debug_device_unlock`.

LMK what you think.

Reply:
The value is now saved on every call to `debug_device_unlock`.